plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt

```diff
@@ -32,13 +34,15 @@ fun abiMetadataProcessor(
             KotlinClassMetadata.transform(header) { metadata ->
                 when (metadata) {
                     is KotlinClassMetadata.Class -> {
-                        metadata.kmClass.removePrivateDeclarations(removeDataClassCopyIfConstructorIsPrivate)
+                        //the single reason for pruning is to leave public class in abi.jar without any content in it
+                        if (pruneClass) metadata.kmClass.visibility = Visibility.PUBLIC
```

Comment:
This line is covered with tests. If removed kotlinc refuses to work with annotation from a private class in an external library. I am not sure it is the best idea, maybe we can just delete this if we only expect those annotations to be read by KSP/javac.

Reply:
A related issue that I found a couple of occurrences of:

```
private class MyClass

@MyAnnotation(MyClass::class.java)
fun publicFun() = ...
```

Again, this is used for annotation processing. My guess is that the goal is to communicate information to an annotation processor without making the classes carrying that information part of the ABI in the end. Would it be simple to trace through annotation values as well and keep any classes referenced from annotations?

Since everything here is under a flag, I would also be OK with landing without keeping annotations and private classes reachable from anntotations. Then we can add a flag for doing that at a later point. For now, we can just not use it (or make things public if that is acceptable for the teams doing this).

Reply:
> we would allow something to compile with ABI jars that would not compile with the full jar
Agree 100%, my bad! Got dragged to a bad decision wanting to write a test somewhat resembling your case. :) Will remove that and look what happens.

> My guess is that the goal is to communicate information to an annotation processor without making the classes carrying that information part of the ABI in the end.

This is a tough one, it is relatively easy to check "typical" places for annotations like methods and classes, not sure about perf impact from that tho (/cc @udalov) but what bothers me is a depth of the rabbit hole =\ @madsager, let's say we figure out that we need to keep `private class MyClass` in the abi.jar what extra information from that class also needs to be kept? It's methods, interfaces, super(recursively?!), inner classes, etc? 

Kotlin allows having private/internal members in annotation arguments no matter what and, I guess, we have to live with it. But what you think is appropriate to keep in such leaked classes? I vote for "nothing" but not sure if it's acceptable. I can see a usecase for that right away: having JSR-330's `@Qualifier` with private class as a key like: `@NoMatterPrivateOrPublicQualifier(PrivateClassAsAKey::class)` so at least we need to keep the private class referenced here. As for other information - I wish we could just make it the problem of the annotation processor author :)

___

@udalov can you please share your thoughts on this one? I think I'm starting to get analysis paralysis here >_<

Reply:
Yes, I agree. I'm leaning towards not trying to fix the annotation issues caused by us doing annotation processing on ABI jars and incorrectly assuming that an annotation processor can inspect non-ABI members in an ABI jar. I think jvm-abi-gen should maintain what is public ABI for compilation with javac and kotlinc. I'm not aware of any cases where javac and kotlinc rely on private annotations or private class constants in annotations for compilation. Which could argue in favor of stripping private annotations from ABI jars and having annotation processor authors who want to use ABI jars deal with the fallout of that (the fallout is small on my end).

Reply:
I agree with Mads about not fixing the annotation issue. If the flag is enabled, private annotation classes should be removed just as other private classes.